### ISSUE TITLE
HTTP/2 & BigPipe

### ISSUE THREAD

**Issue by mgifford** [wet-boew/wet-boew#7404]
There's lots of discussions in the Drupal world about [HTTP/2](https://en.wikipedia.org/wiki/HTTP/2) & [BigPipe](https://www.facebook.com/notes/facebook-engineering/bigpipe-pipelining-web-pages-for-high-performance/389414033919).

Coming from CodefestWxT I know that there has been some informal discussions about HTTP/2 & some of the ideas that are part of BigPipe.

Still, I couldn't find a thread that incorporated either here.

Drupal 8.1 will probably come [bundled with BigPipe](https://www.drupal.org/node/2469431) and it may be the only CMS that comes bundled with this approach to content delivery.

I have asked Facebook's Accessibility team about this and haven't heard back. I don't think it would have the problems that traditional AJAX pages would have though. 

With HTTP/2 a lot of how CSS/JS files are compiled together needs to be [re-thought](https://blog.cloudflare.com/http-2-for-web-developers/). 

So where is the WxT community at looking at these issues?

**Comment by LaurentGoderre** [comment]
I believe that WET might need to be agnostic about it, meaning, making it possible to compile or use not and the implementation would dictate which one should be used. With that being said, there are some issues i cab think about, like potentially loosing the ability to adapt the look of components for each theme

**Comment by LaurentGoderre** [comment]
That is my personal opinion though

**Comment by mgifford** [comment]
Thanks @LaurentGoderre - I really don't know enough either, but this does seem to be the direction that the web will be going.  Only forward looking projects are going to be adopting it at this point, but WxT has played that role for the GoC.  

I don't know if it will be possible to be agnostic about it. So many of the central tenants of web performance are going to be challenged by this. It's a nice goal and maybe you'd be able to build for both.

**Comment by LaurentGoderre** [comment]
It is totally possible to be agnostic by providing the choice. What we can do is minify all the files seperatly, which is the way you would do it with HTTP2 (with HTTP Push...pushing all the plugin files with the core). For thpse who can't or wont use this, we could simply combine all these minified files into a single one.

That way we ate truly agnostic. We have long be shackled by the ability, or lack of ability thereof, of the server. Maybe we can finally move past this and truly take advantage of the server, while still be helpful to those who don't want this.

**Comment by mgifford** [comment]
Agreed with HTTP2 that a bunch of it is tied down to to simply minifying the files separately. However, it's also about only delivering those files that are going to be used for the context. Rather than sending all the CSS/JS that you might use on a page, my understanding is that with HTTP2 you only want to send those bits that are going to be used on that particular page.  If you aren't using a library on a page, it shouldn't be sent to the browser.

With BigPipe though it is different again.  By first sends the cacheable parts of the page and then the dynamic/uncacheable parts there can be a radical reduction in page load times. You flush the initial page first, and then stream the replacements for the placeholders for all subsequent pages you can have substantially faster load times. This requires to re-think how the page is built though and what bits are being implemented.  There's a video for it here https://www.drupal.org/project/big_pipe

**Comment by LaurentGoderre** [comment]
Since WET Core is a front-end farmework, the work towards HTTP2 should be limited to allowing server-side technology, like Drupal, that can know what plugins are on the page to just send those. However, that remains the server-side implementation and if a particular implementation can't do this, there is nothing wrong with sending all of it in individual chunks, it wouldn't be any worse than the current state.

As with BigPipe, I am not familiar enough with it and probably will need to look more into it.

**Comment by mgifford** [comment]
I should have written 2 issues..  Sorry.

And very true "sending all of it in individual chunks, it wouldn't be any worse than the current state." - hopefully though a customized back-end could reduce the number of files that need to be sent. @sylus have you taken a look at either HTTP2 or BigPipe?

**Comment by LaurentGoderre** [comment]
In case it wasn't clear, I am very excited to look at the opportunity these new techs given! ;)

**Comment by mgifford** [comment]
Cool..  I thought as much. Now if we can just interest the rest of the GoC.

**Comment by hsrudnicki** [comment]
Does this remain an issue or otherwise? Please update the status of this, whether this can be closed or otherwise? Thx

**Comment by mgifford** [comment]
These might be beyond the scope of WxT although they do definitely impact it. 

I guess we could close it if there isn't interest in these performance technologies.

**Comment by hsrudnicki** [comment]
Based on the above, this can be closed. If you think otherwise or have an interest in the performance technologies please re-open it or make the status update of this.

**Comment by LaurentGoderre** [comment]
This is still of value for next versions
